<?xml version="1.0"?>

<!-- Inertia Calculating; -->
<!-- http://wiki.ros.org/urdf/Tutorials/Adding%20Physical%20and%20Collision%20Properties%20to%20a%20URDF%20Model-->
<!-- https://en.wikipedia.org/wiki/List_of_moments_of_inertia-->


<robot name="robotic_arm">

    <link name="world"/>
      <link name="base_link">
          <visual>
              <geometry>
                  <cylinder length="0.05" radius="0.2"/>
              </geometry>
              <material name="Black">
                  <color rgba="0 0 0 1"/>
              </material>
              <origin rpy="0 0 0" xyz="0 0 0.025"/>
          </visual>
  
      <collision>
          <geometry>
              <cylinder length="0.05" radius="0.2"/>
          </geometry>
          <origin rpy="0 0 0" xyz="0 0 0.025"/>
      </collision>
  
      <inertial>
          <origin rpy="0 0 0" xyz="0 0 0.025"/>
          <mass value="5.0"/>
          <inertia ixx="0.051" ixy="0.0" ixz="0.0" iyy="0.051" iyz="0.0" izz="0.05"/>
      </inertial>
      
  
      </link>
  
    <joint name="fixed" type="fixed">
      <parent link="world"/>
      <child link="base_link"/>
      <dynamics damping="10" friction="1.0"/>
    </joint>
  
    <link name="link_1">
      <visual>
          <geometry>
              <cylinder length="0.5" radius="0.08"/>
          </geometry>
          <material name="blue">
              <color rgba="0 0 0.8 1"/>
            </material>
          <origin rpy="0 0 0" xyz="0 0 0.25"/>
      </visual>
  
      <collision>
          <geometry>
              <cylinder length="0.5" radius="0.08"/>
          </geometry>
          <origin rpy="0 0 0" xyz="0 0 0.25"/>
      </collision>
      
      <inertial>
          <origin rpy="0 0 0" xyz="0 0 0.25"/>
          <mass value="5.0"/>
          <inertia ixx="0.112" ixy="0.0" ixz="0.0" iyy="0.112" iyz="0.0" izz="0.008"/>
      </inertial>
  
    </link>
    
    <joint name="joint_1" type="revolute">
      <axis xyz="0 0 1"/>
      <parent link="base_link"/>
      <child link="link_1"/>
      <origin rpy="0 0 0" xyz="0.0 0.0 0.05"/>
      <limit lower="-3.34" upper="3.34" effort="10" velocity="0.5"/>
      <dynamics damping="10" friction="1.0"/>
    </joint>
  
  <link name="link_2">
      <inertial>
          <origin rpy="0 0 0" xyz="0 0 0.2"/>
          <mass value="2.0"/>
          <inertia ixx="0.0004" ixy="0.0" ixz="0.0" iyy="0.0004" iyz="0.0" izz="0.0006"/>
      </inertial>
  
      <visual>
          <geometry>
          <cylinder length="0.1" radius="0.08"/>
          </geometry>
          <material name="Red">
              <color rgba="1 0 0 1"/>
            </material>
      </visual>
  
      <collision>
          <geometry>
              <cylinder length="0.1" radius="0.08"/>
          </geometry>
      </collision>
  
  </link>
  
  <joint name="joint_2" type="revolute">
  <axis xyz="0 0 1"/>
  <parent link="link_1"/>
  <child link="link_2"/>
  <origin rpy="0 1.5708 0" xyz="0.0 -0.005 0.58"/>
  <limit lower="-0.25" upper="3.34" effort="10" velocity="0.5"/>
  <dynamics damping="10" friction="1.0"/>
  </joint>
  
  <link name="link_3">
      <inertial>
          <origin rpy="0 0 0" xyz="0 0 0.2"/>
          <mass value="1.0"/>
          <inertia ixx="0.013" ixy="0.0" ixz="0.0" iyy="0.13" iyz="0.0" izz="0.00025"/>
      </inertial>
      
      <visual>
          <geometry>
          <cylinder length="0.4" radius="0.05"/>
          </geometry>
          <material name="blue">
              <color rgba="0.3 0.8 0.5 1"/>
            </material>
          </visual>
  
      <collision>
          <geometry>
              <cylinder length="0.4" radius="0.05"/>
          </geometry>
      </collision>
  </link>
  
  <joint name="joint_3" type="fixed">
  <parent link="link_2"/>
  <child link="link_3"/>
  <origin rpy="1.57 0 0" xyz="0.0 0.2 0 "/>
  <dynamics damping="10" friction="1.0"/>
  </joint>
  
  <link name="link_4">
      <inertial>
          <origin rpy="0 0 0" xyz="0 0 0.2"/>
          <mass value="1"/>
          <inertia ixx="0.0017" ixy="0.0" ixz="0.0" iyy="0.0017" iyz="0.0" izz="0.00025"/>
      </inertial>
  
      <visual>
          <geometry>
          <cylinder length="0.1" radius="0.06"/>
          </geometry>
          <material name="Red">
              <color rgba="1 0 0 1"/>
            </material>
      </visual>
  
      <collision>
          <geometry>
              <cylinder length="0.1" radius="0.06"/>
          </geometry>
      </collision>
  
  </link>
  
  <joint name="joint_4" type="revolute">
  <parent link="link_3"/>
  <child link="link_4"/>
  <origin rpy="1.57 0 0" xyz=" 0 0 -0.25"/>
  <axis xyz=" 0 0 1"/>
  <limit lower="-1.92" upper="1.92" effort="10" velocity="0.5"/>
  <dynamics damping="10" friction="1.0"/>
  </joint>
  
  
  
  
  <link name="link_5">
      <inertial>
          <origin rpy="0 0 0" xyz="0 0 0.2"/>
          <mass value="0.5"/>
          <inertia ixx="0.0038625" ixy="0.0" ixz="0.0" iyy="0.0038625" iyz="0.0" izz="0.0000375"/>
      </inertial>
      
      <visual>
          <geometry>
          <cylinder length="0.3" radius="0.03"/>
          </geometry>
          <material name="yellow">
              <color rgba="0 1 0.5 1"/>
          </material>
          </visual>
  
      <collision>
          <geometry>
              <cylinder length="0.3" radius="0.03"/>
          </geometry>
      </collision>
  </link>
  
  <joint name="joint_5" type="fixed">
  <parent link="link_4"/>
  <child link="link_5"/>
  <origin rpy="1.57 0 0" xyz="0.0 -0.2 0 "/>
  <dynamics damping="10" friction="1.0"/>
  
  </joint>
  
  
  
<gazebo reference="base_link">
    <material>Gazebo/Black</material>
</gazebo>

<gazebo reference="link_1">
    <material>Gazebo/Blue</material>
</gazebo>

<gazebo reference="link_3">
    <material>Gazebo/Blue</material>
</gazebo>

<gazebo reference="link_2">
    <material>Gazebo/Red</material>
</gazebo>

<gazebo reference="link_4">
    <material>Gazebo/Red</material>
</gazebo>

<gazebo reference="link_5">
    <material>Gazebo/Orange</material>
</gazebo>

  
  
  <!-- Ros2 Control Interface  -->
  

    
    
    
  
    <ros2_control name="GazeboSystem" type="system">
        <hardware>
          <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>
    
    
    <joint name="joint_1">
        <command_interface name="position">
          <param name="min">-3.14</param>
          <param name="max">3.14</param>
        </command_interface>
        <state_interface name="position"/>
        <param name="initial_position">0.0</param>  
      </joint>
      <joint name="joint_2">
        <command_interface name="position">
          <param name="min">-3.14</param>
          <param name="max">3.14</param>
        </command_interface>
        <state_interface name="position"/>
        <param name="initial_position">-1.57</param>  
      </joint>
      <joint name="joint_4">
        <command_interface name="position">
          <param name="min">-3.14</param>
          <param name="max">3.14</param>
        </command_interface>
        <state_interface name="position"/>
        <param name="initial_position">0.0</param>  
      </joint>
    
    </ros2_control>

    
  
    <gazebo>
    <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
      <robot_param>robot_description</robot_param>
      <robot_param_node>robot_state_publisher</robot_param_node>
      <parameters>$(find robotic_arm)/config/jtc.yaml</parameters>
    </plugin>
    </gazebo>

    

  </robot>
